google/cloud/completion_queue_test.cc

```diff
@@ -488,6 +489,134 @@ TEST(CompletionQueueTest, RunAsyncThread) {
   for (auto& t : runners) t.join();
 }
 
+TEST(CompletionQueueTest, RunAsyncParallel) {
+  auto impl = std::make_shared<internal::DefaultCompletionQueueImpl>();
+  CompletionQueue cq(impl);
+  auto constexpr kRunners = 16;
+  std::vector<std::thread> runners(kRunners);
+  std::generate(runners.begin(), runners.end(),
+                [&cq] { return std::thread{[&cq] { cq.Run(); }}; });
+
+  std::mutex mu;
+  std::condition_variable cv;
+  std::deque<promise<void>> waiting;
+  std::size_t actual_max_runners = 0;
+  auto add_promise = [&] {
+    promise<void> p;
+    auto f = p.get_future();
+    std::unique_lock<std::mutex> lk(mu);
+    waiting.push_back(std::move(p));
+    actual_max_runners = (std::max)(actual_max_runners, waiting.size());
+    lk.unlock();
+    cv.notify_one();
+    return f;
+  };
+  auto wait = [&] {
+    std::unique_lock<std::mutex> lk(mu);
+    cv.wait(lk, [&] { return !waiting.empty(); });
+    auto f = std::move(waiting.front());
+    waiting.pop_front();
+    return f;
+  };
+
+  // Start an async function and block it... this will help us verify that
+  // different functions get called in different threads. Start with just one:
+  cq.RunAsync([&] { add_promise().get(); });
+  wait().set_value();
+
+  // Verify we never exceed 15 threads running in parallel, because we want to
+  // reserve 1 thread for the I/O loop.
+  auto constexpr kIterations = 100;
+  auto constexpr kRepeats = 8;
+  for (int i = 0; i != kIterations; ++i) {
+    auto const n = kRepeats * kRunners;
+    for (int j = 0; j != n; ++j) cq.RunAsync([&] { add_promise().get(); });
+    for (int j = 0; j != n; ++j) wait().set_value();
+  }
+  EXPECT_GT(kRunners, actual_max_runners);
+
+  cq.Shutdown();
+  for (auto& t : runners) t.join();
+  // We do not expect a lot of calls to Notify(), maybe kRunners * kIterations
+  // at most, but let's be conservative.
+  EXPECT_LE(impl->notify_counter(), kIterations * kRunners * kRepeats / 4);
+}
+
+TEST(CompletionQueueTest, RunAsyncSingleThreaded) {
+  auto impl = std::make_shared<internal::DefaultCompletionQueueImpl>();
+  CompletionQueue cq(impl);
+  std::thread t{[&] { cq.Run(); }};
+
+  // Make each function in RunAsync() block under control of this thread.
+  std::mutex mu;
+  std::condition_variable cv;
+  std::deque<promise<void>> waiting;
+  auto add_promise = [&] {
+    std::unique_lock<std::mutex> lk(mu);
+    waiting.emplace_back();
+    auto f = waiting.back().get_future();
```

Comment:
Any reason this differs from the pushes to `waiting` in the surrounding tests?

Reply:
None, fixed by refactoring.